Illegal instant due to time zone offset transition

Hi all,

I have the following query -
```
{
	"dataSource": "player-events-historical",
	"intervals": ["2017-10-10T00:00:00-03:00\/2017-11-09T23:59:59-03:00"],
	"granularity": {
		"type": "period",
		"period": "P1D",
		"timeZone": "America\/Sao_Paulo"
	},
	"aggregations": [{
		"type": "longSum",
		"name": "count",
		"fieldName": "count"
	}],
	"queryType": "timeseries"
}
```
When I post it to Druid, I'm getting the following exception -
```
# curl -HContent-Type:application/json -i -d @/tmp/sp.json localhost:8082/druid/v2/
HTTP/1.1 500 Internal Server Error
Date: Mon, 01 Jan 2018 15:44:02 GMT
Content-Type: application/json
Content-Length: 321
Server: Jetty(9.3.16.v20170120)

{"error":"Unknown exception","errorMessage":"Value 0 for hourOfDay is not supported: Illegal instant due to time zone offset transition (daylight savings time 'gap'): 2017-10-15T00:00:00.000 (America/Sao_Paulo)","errorClass":"org.joda.time.IllegalFieldValueException","host":"ip-..."}
```
If I change the timezone to `Etc/GMT+3` it works fine. 
I see the following exception in one of the historical nodes -
```
2018-01-01T15:44:02,182 ERROR [processing-12] io.druid.query.ChainedExecutionQueryRunner - Exception with one ofthe sequences!
org.joda.time.IllegalFieldValueException: Value 0 for hourOfDay is not supported: Illegal instant due to time zone offset transition (daylight savings time 'gap'): 2017-10-15T00:00:00.000 (America/Sao_Paulo)
        at org.joda.time.chrono.ZonedChronology$ZonedDateTimeField.set(ZonedChronology.java:486) ~[joda-time-2.8.2.jar:2.8.2]
        at io.druid.java.util.common.granularity.PeriodGranularity.truncate(PeriodGranularity.java:314) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.granularity.PeriodGranularity.bucketStart(PeriodGranularity.java:124) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.granularity.Granularity.bucketEnd(Granularity.java:124) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.segment.QueryableIndexStorageAdapter.makeCursors(QueryableIndexStorageAdapter.java:213) ~[druid-processing-0.10.0.jar:0.10.0]
        at io.druid.query.QueryRunnerHelper.makeCursorBasedQuery(QueryRunnerHelper.java:61) ~[druid-processing-0.10.0.jar:0.10.0]
        at io.druid.query.timeseries.TimeseriesQueryEngine.process(TimeseriesQueryEngine.java:50) ~[druid-processing-0.10.0.jar:0.10.0]
        at io.druid.query.timeseries.TimeseriesQueryRunnerFactory$TimeseriesQueryRunner.run(TimeseriesQueryRunnerFactory.java:102) ~[druid-processing-0.10.0.jar:0.10.0]
        at io.druid.query.ReferenceCountingSegmentQueryRunner.run(ReferenceCountingSegmentQueryRunner.java:55) ~[druid-processing-0.10.0.jar:0.10.0]
        at io.druid.query.MetricsEmittingQueryRunner$2.get(MetricsEmittingQueryRunner.java:102) ~[druid-processing-0.10.0.jar:0.10.0]
        at io.druid.query.MetricsEmittingQueryRunner$2.get(MetricsEmittingQueryRunner.java:98) ~[druid-processing-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.guava.LazySequence.accumulate(LazySequence.java:40) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.guava.WrappingSequence$1.get(WrappingSequence.java:50) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.guava.SequenceWrapper.wrap(SequenceWrapper.java:55) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.guava.WrappingSequence.accumulate(WrappingSequence.java:45) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.guava.Sequences.toList(Sequences.java:150) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.query.BySegmentQueryRunner.run(BySegmentQueryRunner.java:56) ~[druid-processing-0.10.0.jar:0.10.0]
        at io.druid.query.MetricsEmittingQueryRunner$2.get(MetricsEmittingQueryRunner.java:102) ~[druid-processing-0.10.0.jar:0.10.0]
        at io.druid.query.MetricsEmittingQueryRunner$2.get(MetricsEmittingQueryRunner.java:98) ~[druid-processing-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.guava.LazySequence.accumulate(LazySequence.java:40) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.guava.WrappingSequence$1.get(WrappingSequence.java:50) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.guava.SequenceWrapper.wrap(SequenceWrapper.java:55) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.guava.WrappingSequence.accumulate(WrappingSequence.java:45) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.query.spec.SpecificSegmentQueryRunner$2.accumulate(SpecificSegmentQueryRunner.java:83) ~[druid-processing-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.guava.WrappingSequence$1.get(WrappingSequence.java:50) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.query.spec.SpecificSegmentQueryRunner.doNamed(SpecificSegmentQueryRunner.java:169) ~[druid-processing-0.10.0.jar:0.10.0]
        at io.druid.query.spec.SpecificSegmentQueryRunner.access$200(SpecificSegmentQueryRunner.java:43) ~[druid-processing-0.10.0.jar:0.10.0]
        at io.druid.query.spec.SpecificSegmentQueryRunner$3.wrap(SpecificSegmentQueryRunner.java:149) ~[druid-processing-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.guava.WrappingSequence.accumulate(WrappingSequence.java:45) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.guava.WrappingSequence$1.get(WrappingSequence.java:50) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.query.CPUTimeMetricQueryRunner$1.wrap(CPUTimeMetricQueryRunner.java:78) ~[druid-processing-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.guava.WrappingSequence.accumulate(WrappingSequence.java:45) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.java.util.common.guava.Sequences.toList(Sequences.java:150) ~[java-util-0.10.0.jar:0.10.0]
        at io.druid.query.ChainedExecutionQueryRunner$1$1$1.call(ChainedExecutionQueryRunner.java:129) [druid-processing-0.10.0.jar:0.10.0]
        at io.druid.query.ChainedExecutionQueryRunner$1$1$1.call(ChainedExecutionQueryRunner.java:119) [druid-processing-0.10.0.jar:0.10.0]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_131]
        at io.druid.query.PrioritizedListenableFutureTask.run(PrioritizedExecutorService.java:271) [druid-processing-0.10.0.jar:0.10.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_131]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_131]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
Caused by: org.joda.time.IllegalInstantException: Illegal instant due to time zone offset transition (daylight savings time 'gap'): 2017-10-15T00:00:00.000 (America/Sao_Paulo)
        at org.joda.time.chrono.ZonedChronology$ZonedDateTimeField.set(ZonedChronology.java:485) ~[joda-time-2.8.2.jar:2.8.2]
        ... 39 more
```
Since the failed timestamp occurs somewhere within the requested interval (it's not close to the edges), I believe this is a bug in Druid (I'm using Druid 0.10.0), but please let me know if you think I'm doing something wrong.

Thank you!

Eran

--- a/core/src/main/java/org/apache/druid/java/util/common/granularity/PeriodGranularity.java
+++ b/core/src/main/java/org/apache/druid/java/util/common/granularity/PeriodGranularity.java
@@ -307,8 +307,7 @@ public class PeriodGranularity extends Granularity implements JsonSerializable
         }
         return t;
       } else {
-        t = chronology.hourOfDay().roundFloor(t);
-        return chronology.hourOfDay().set(t, 0);
+        return chronology.dayOfMonth().roundFloor(t);
       }
     }
 
@@ -330,8 +329,7 @@ public class PeriodGranularity extends Granularity implements JsonSerializable
         }
         return t;
       } else {
-        t = chronology.minuteOfHour().roundFloor(t);
-        return chronology.minuteOfHour().set(t, 0);
+        return chronology.hourOfDay().roundFloor(t);
       }
     }
 
@@ -350,8 +348,7 @@ public class PeriodGranularity extends Granularity implements JsonSerializable
         }
         return t;
       } else {
-        t = chronology.secondOfMinute().roundFloor(t);
-        return chronology.secondOfMinute().set(t, 0);
+        return chronology.minuteOfHour().roundFloor(t);
       }
     }
 

--- a/core/src/test/java/org/apache/druid/java/util/common/GranularityTest.java
+++ b/core/src/test/java/org/apache/druid/java/util/common/GranularityTest.java
@@ -781,7 +781,7 @@ public class GranularityTest
     // Sao Paulo daylight savings time in 2017 starts at midnight. When we spring forward, 00:00:00 doesn't exist.
     // (The clock goes straight from 23:59:59 to 01:00:00.) This test verifies we handle the case correctly while
     // iterating through Paulistano days.
-    final DateTimeZone saoPaulo = DateTimeZone.forID("America/Sao_Paulo");
+    final DateTimeZone saoPaulo = DateTimes.inferTzFromString("America/Sao_Paulo");
     final PeriodGranularity granSaoPauloDay = new PeriodGranularity(
         Period.days(1),
         null,

--- a/core/src/test/java/org/apache/druid/java/util/common/GranularityTest.java
+++ b/core/src/test/java/org/apache/druid/java/util/common/GranularityTest.java
@@ -19,6 +19,7 @@
 
 package org.apache.druid.java.util.common;
 
+import com.google.common.collect.ImmutableList;
 import org.apache.druid.java.util.common.granularity.Granularities;
 import org.apache.druid.java.util.common.granularity.Granularity;
 import org.apache.druid.java.util.common.granularity.GranularityType;
@@ -34,6 +35,8 @@ import org.junit.Test;
 
 import java.util.Iterator;
 import java.util.NoSuchElementException;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
 
 public class GranularityTest
 {
@@ -770,6 +773,40 @@ public class GranularityTest
     catch (IAE e) {
       // pass
     }
+  }
+
+  @Test // Regression test for https://github.com/apache/incubator-druid/issues/5200.
+  public void testIncrementOverSpringForward()
+  {
+    // Sao Paulo daylight savings time in 2017 starts at midnight. When we spring forward, 00:00:00 doesn't exist.
+    // (The clock goes straight from 23:59:59 to 01:00:00.) This test verifies we handle the case correctly while
+    // iterating through Paulistano days.
+    final DateTimeZone saoPaulo = DateTimeZone.forID("America/Sao_Paulo");
+    final PeriodGranularity granSaoPauloDay = new PeriodGranularity(
+        Period.days(1),
+        null,
+        saoPaulo
+    );
+
+    final Iterable<Interval> intervals = granSaoPauloDay.getIterable(
+        new Interval(
+            new DateTime("2017-10-14", saoPaulo),
+            new DateTime("2017-10-17", saoPaulo)
+        )
+    );
+
+    // Similar to what query engines do: call granularity.bucketStart on the datetimes returned by their cursors.
+    // (And the cursors, in turn, use getIterable like above.)
+    Assert.assertEquals(
+        ImmutableList.of(
+            new DateTime("2017-10-14", saoPaulo),
+            new DateTime("2017-10-15T01", saoPaulo),
+            new DateTime("2017-10-16", saoPaulo)
+        ),
+        StreamSupport.stream(intervals.spliterator(), false)
+                     .map(interval -> granSaoPauloDay.bucketStart(interval.getStart()))
+                     .collect(Collectors.toList())
+    );
   }
 
   private static class PathDate

